pytorch_lightning/callbacks/finetuning.py

```diff
@@ -0,0 +1,221 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+r"""
+Finetunning Callback
+^^^^^^^^^^^^^^^^^^^^
+Freeze and unfreeze models for finetunning purposes
+"""
+from typing import Callable, Generator, Optional
+
+import torch
+from torch.nn import Module
+from torch.nn.modules.container import Sequential
+from torch.optim.optimizer import Optimizer
+
+from pytorch_lightning import _logger as log
+from pytorch_lightning.callbacks.base import Callback
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+BN_TYPES = (torch.nn.BatchNorm1d, torch.nn.BatchNorm2d, torch.nn.BatchNorm3d)
+
+
+def _make_trainable(module: Module) -> None:
+    """Unfreezes a given module.
+    Args:
+        module: The module to unfreeze
+    """
+    for param in module.parameters():
+        param.requires_grad = True
+    module.train()
+
+
+def _recursive_freeze(module: Module,
+                      train_bn: bool = True) -> None:
+    """Freezes the layers of a given module.
+    Args:
+        module: The module to freeze
+        train_bn: If True, leave the BatchNorm layers in training mode
+    """
+    children = list(module.children())
+    if not children:
+        if not (isinstance(module, BN_TYPES) and train_bn):
+            for param in module.parameters():
+                param.requires_grad = False
+            module.eval()
+        else:
+            # Make the BN layers trainable
+            _make_trainable(module)
+    else:
+        for child in children:
+            _recursive_freeze(module=child, train_bn=train_bn)
+
+
+def filter_params(module: Module,
+                  train_bn: bool = True) -> Generator:
+    """Yields the trainable parameters of a given module.
+    Args:
+        module: A given module
+        train_bn: If True, leave the BatchNorm layers in training mode
+    Returns:
+        Generator
+    """
+    children = list(module.children())
+    if not children:
+        if not (isinstance(module, BN_TYPES) and train_bn):
+            for param in module.parameters():
+                if param.requires_grad:
+                    yield param
+    else:
+        for child in children:
+            for param in filter_params(module=child, train_bn=train_bn):
+                yield param
+
+
+def multiplicative(epoch):
+    return 2
+
+
+class BaseFinetuningCallback(Callback):
+
+    r"""
+    BaseFinetuningCallback.
+    Overrides the finetunning_function and add your own logic there.
```

Comment:
```suggestion
    Override the finetunning_function and add your own logic there.
```

Can you also mention that it is done `on_epoch_start`?